File: services/taiwan-architect-kb-mcp/src/store.py
"""JSON-file-backed knowledge base store, stdlib-only."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

DEFAULT_DATA_PATH = Path(__file__).resolve().parent.parent / "data" / "building_codes.json"


@dataclass(frozen=True)
class KbEntry:
    id: str
    category: str
    title: str
    summary: str
    full_text: str
    status: str
    source_url: str
    tags: list[str] = field(default_factory=list)

    def searchable_text(self) -> str:
        return " ".join([self.title, self.summary, self.full_text, " ".join(self.tags)])


class KnowledgeBaseStore:
    def __init__(self, data_path: Path | None = None):
        self._data_path = data_path or DEFAULT_DATA_PATH
        self._entries: dict[str, KbEntry] = {}
        self._load()

    def _load(self) -> None:
        raw = json.loads(self._data_path.read_text(encoding="utf-8"))
        for item in raw.get("entries", []):
            entry = KbEntry(
                id=item["id"],
                category=item.get("category", ""),
                title=item.get("title", ""),
                summary=item.get("summary", ""),
                full_text=item.get("full_text", ""),
                status=item.get("status", ""),
                source_url=item.get("source_url", ""),
                tags=list(item.get("tags", [])),
            )
            self._entries[entry.id] = entry

    def get(self, entry_id: str) -> KbEntry | None:
        return self._entries.get(entry_id)

    def all(self) -> list[KbEntry]:
        return list(self._entries.values())

    def search(self, query: str, category: str | None = None) -> list[KbEntry]:
        """Naive substring/keyword search over title, summary, full_text, tags."""
        terms = [t for t in query.replace("　", " ").split() if t] or [query]
        results = []
        for entry in self._entries.values():
            if category and entry.category != category:
                continue
            haystack = entry.searchable_text()
            if any(term in haystack for term in terms):
                results.append(entry)
        return results

File: services/taiwan-architect-kb-mcp-cloudflare/src/lib/store.js
/**
 * In-memory knowledge base store, zero dependencies.
 *
 * Takes the parsed entries array directly (dependency injection) rather
 * than reading a file itself, so the same code works both in a Cloudflare
 * Worker (where the JSON is bundled via `import data from "../data/..."`)
 * and under plain Node for local testing (where it's loaded with
 * `JSON.parse(fs.readFileSync(...))`).
 */

export class KnowledgeBaseStore {
  /** @param {Array<Record<string, any>>} entries */
  constructor(entries) {
    /** @type {Map<string, Record<string, any>>} */
    this._entries = new Map();
    for (const item of entries) {
      this._entries.set(item.id, {
        id: item.id,
        category: item.category ?? "",
        title: item.title ?? "",
        summary: item.summary ?? "",
        fullText: item.full_text ?? "",
        status: item.status ?? "",
        sourceUrl: item.source_url ?? "",
        tags: item.tags ?? [],
      });
    }
  }

  _searchableText(entry) {
    return [entry.title, entry.summary, entry.fullText, ...entry.tags].join(" ");
  }

  get(entryId) {
    return this._entries.get(entryId) ?? null;
  }

  all() {
    return [...this._entries.values()];
  }

  /** @param {string} query @param {string} [category] */
  search(query, category) {
    const terms = query.split(/\s+/).filter(Boolean);
    const effectiveTerms = terms.length > 0 ? terms : [query];
    const results = [];
    for (const entry of this._entries.values()) {
      if (category && entry.category !== category) continue;
      const haystack = this._searchableText(entry);
      if (effectiveTerms.some((term) => haystack.includes(term))) {
        results.push(entry);
      }
    }
    return results;
  }
}
